kaolin/ops/spc/spc.py

```diff
@@ -0,0 +1,399 @@
+#
+# Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.
+#
+# NVIDIA CORPORATION and its licensors retain all intellectual property
+# and proprietary rights in and to this software, related documentation
+# and any modifications thereto.  Any use, reproduction, disclosure or
+# distribution of this software and related documentation without an express
+# lice
+# nse agreement from NVIDIA CORPORATION is strictly prohibited.
+#
+
+import math
+from torch import nn
+from torch.autograd import Function
+import torch
+import numpy as np
+
+from kaolin import _C
+
+
+
+def ScanOctrees(octrees, olens):
+    """
+    Scan batch of octree tensor.
+
+    Args:
+        octrees (torch.Tensor).dtype(torch::kByte): Shape = (b, k)
+        olens (torch.Tensor).dtype(torch::kInt32): Shape = (b)
+
+    Returns:
+        (torch.Tensor).dtype(torch::kInt32):
+            Tensor of corresponding point lengths. Shape = (b)
+        (torch.Tensor).dtype(torch::kInt32):
+            Result of the exclusive sum of octree bytes bit counts, per batch. Shape = (b, k+b)
+    """    
+    return _C.ops.spc.ScanOctrees(octrees, olens)
+
+def DecodePoints(octrees, olens, plens, exsum):
+    return _C.ops.spc.DecodePoints(octrees, olens, plens, exsum)
+
+# def SetState(octrees, points, pyramid, exsum):
+#     return _C.ops.spc.SetState(octrees, points, pyramid, exsum)
+
+# def SetState(octrees, olens):
+#     plens, exsum = ScanOctrees(octrees, olens)
+#     points, pyramid = DecodePoints(octrees, olens, plens, exsum)
+#     return _C.ops.spc.SetState(octrees, olens, points, plens, pyramid, exsum)
+
+
+
+
+#  ####### Convolution ######################################
+
+
+class Conv3dFunction(Function):
+    @staticmethod
+    def forward(ctx, octrees, points, pyramid, exsum, inputs, params, kernel_vectors, jump):
+        if not inputs.is_contiguous():
+            inputs = inputs.contiguous()
+
+        outputs = inputs.new()
+        _C.ops.spc.Conv3d_forward(octrees, points, pyramid, exsum, inputs, outputs, params, kernel_vectors, jump)
+
+        ctx.octrees = octrees
+        ctx.points = points
+        ctx.pyramid = pyramid
+        ctx.exsum = exsum
+        ctx.inputs = inputs
+        ctx.params = params
+        ctx.kernel_vectors = kernel_vectors
+        ctx.jump = jump
+
+        return outputs
+
+    @staticmethod
+    def backward(ctx, grad_outputs):
+        if not grad_outputs.is_contiguous():
+            grad_outputs = grad_outputs.contiguous()
+
+        d_inputs = grad_outputs.new()
+        d_params = grad_outputs.new()  # this seems wrong!
+
+        _C.ops.spc.Conv3d_backward(ctx.octrees, ctx.points, ctx.pyramid, ctx.exsum, ctx.inputs, d_inputs,
+                                 grad_outputs,
+                                 ctx.params, d_params,
+                                 ctx.kernel_vectors, ctx.jump )
+
+        return None, None, None, None, d_inputs, d_params, None, None
+
+
+class Conv3d(nn.Module):
```

Comment:
Can you add a test case or an example of usage of the Conv3d?

Reply:
I added some meager test code in  test/pythoin/kaolin/spc. Needs improvement.